flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionSlotAllocator.java

```diff
@@ -0,0 +1,248 @@
+	}
+
+	@Override
+	public void cancel(ExecutionVertexID executionVertexId) {
+		SlotExecutionVertexAssignment slotExecutionVertexAssignment = pendingSlotAssignments.get(executionVertexId);
+		if (slotExecutionVertexAssignment != null) {
+			slotExecutionVertexAssignment.getLogicalSlotFuture().cancel(false);
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> stop() {
+		List<ExecutionVertexID> executionVertexIds = new ArrayList<>(pendingSlotAssignments.keySet());
+		executionVertexIds.stream().forEach(executionVertexID -> cancel(executionVertexID));
+
+		return CompletableFuture.completedFuture(null);
+	}
+
+	/**
+	 * Calculates the preferred locations for an execution.
+	 * It will first try to use preferred locations based on state,
+	 * if null, will use the preferred locations based on inputs.
+	 */
+	private static CompletableFuture<Collection<TaskManagerLocation>> calculatePreferredLocations(
+			ExecutionVertexID executionVertexId,
+			Collection<TaskManagerLocation> preferredLocationsBasedOnState,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+
+		if (!preferredLocationsBasedOnState.isEmpty()) {
+			return CompletableFuture.completedFuture(preferredLocationsBasedOnState);
+		}
+
+		return getPreferredLocationsBasedOnInputs(executionVertexId, inputsLocationsRetriever);
+	}
+
+	/**
+	 * Gets the location preferences of the execution, as determined by the locations
+	 * of the predecessors from which it receives input data.
+	 * If there are more than {@link ExecutionVertex#MAX_DISTINCT_LOCATIONS_TO_CONSIDER} different locations of source data,
+	 * or neither the sources have not been started nor will be started with the execution together,
+	 * this method returns an empty collection to indicate no location preference.
+	 *
+	 * @return The preferred locations based in input streams, or an empty iterable,
+	 *         if there is no input-based preference.
+	 */
+	@VisibleForTesting
+	static CompletableFuture<Collection<TaskManagerLocation>> getPreferredLocationsBasedOnInputs(
+			ExecutionVertexID executionVertexId,
+			InputsLocationsRetriever inputsLocationsRetriever) {
+		CompletableFuture<Collection<TaskManagerLocation>> preferredLocations =
+				CompletableFuture.completedFuture(Collections.emptyList());
+
+		Collection<CompletableFuture<TaskManagerLocation>> locationsFutures = new ArrayList<>();
+
+		Collection<Collection<ExecutionVertexID>> allProducers =
+				inputsLocationsRetriever.getConsumedResultPartitionsProducers(executionVertexId);
+		for (Collection<ExecutionVertexID> producers : allProducers) {
+
+			for (ExecutionVertexID producer : producers) {
+				Optional<CompletableFuture<TaskManagerLocation>> optionalLocationFuture =
+						inputsLocationsRetriever.getTaskManagerLocation(producer);
+				optionalLocationFuture.ifPresent(
+						(locationFuture) -> locationsFutures.add(locationFuture));
+				// If the parallelism is large, wait for all futures coming back may cost a long time.
```

Comment:
Is this main reason or is it because of the data portion which becomes smaller the larger the number of different locations becomes? Have we seen that waiting for all futures becomes a problem somewhere?

Reply:
I think this is the main reason. As user will change the parallelism according to the total size of data to keep the size of each partition stable. We didn't see the case waiting for a long time since the logic is same in legacy code and for almost all jobs in alibaba, the parallelism is much larger than 8.